File: app/src/main/java/com/dss886/transmis/listen/sms/SmsListener.java
package com.dss886.transmis.listen.sms;

import android.content.BroadcastReceiver;
import android.content.Context;
import android.content.Intent;
import android.os.Bundle;
import android.telephony.SmsMessage;
import android.text.TextUtils;

import com.dss886.transmis.R;
import com.dss886.transmis.base.App;
import com.dss886.transmis.nofity.DingDingSender;
import com.dss886.transmis.nofity.IftttWebhooksSender;
import com.dss886.transmis.nofity.MailGunSender;
import com.dss886.transmis.nofity.MailSender;
import com.dss886.transmis.nofity.TelegramSender;
import com.dss886.transmis.utils.Logger;
import com.dss886.transmis.utils.Settings;
import com.dss886.transmis.utils.StringUtils;
import com.dss886.transmis.utils.Tags;

import java.util.List;
import java.util.Locale;

/**
 * Created by dss886 on 2017/6/29.
 */

public class SmsListener extends BroadcastReceiver {

    private static final String ACTION_SMS_RECEIVED = "android.provider.Telephony.SMS_RECEIVED";

    @Override
    public void onReceive(Context context, Intent intent) {
        Logger.d("SMS Received.");
        if (!Settings.is(Tags.SP_GLOBAL_ENABLE, false)) {
            Logger.d("SMS Transmis has been disable!");
            return;
        }
        if (ACTION_SMS_RECEIVED.equals(intent.getAction())) {
            Bundle bundle = intent.getExtras();
            if (bundle != null) {
                try {
                    Object[] pdus = (Object[]) bundle.get("pdus");
                    assert pdus != null;
                    SmsMessage[] messages = new SmsMessage[pdus.length];
                    for (int i = 0; i < messages.length; i++) {
                        messages[i] = SmsMessage.createFromPdu((byte[]) pdus[i]);
                    }
                    Logger.d("Try To Notify.");
                    doNotify(messages);
                } catch (Exception e) {
                    Logger.e(e.getMessage());
                    e.printStackTrace();
                }
            }
        }
    }

    private void doNotify(SmsMessage[] messages) {
        if (messages == null || messages.length == 0) {
            return;
        }
        String titleRegex = App.sp.getString(Tags.SP_SMS_TITLE_REGEX, App.me().getString(R.string.sms_title_default));
        String contentRegex = App.sp.getString(Tags.SP_SMS_CONTENT_REGEX, App.me().getString(R.string.sms_content_default));
        String content;
        StringBuilder sb = new StringBuilder();
        if (App.sp.getBoolean(Tags.SP_SMS_MERGE_LONG_TEXT, true)) {
            for (SmsMessage message : messages) {
                sb.append(message.getMessageBody());
            }
            String callNumber = messages[0].getOriginatingAddress();
            String smsContent = sb.toString();
            if (tryFilter(callNumber, smsContent)) {
                return;
            }
            content = String.format(Locale.CHINA, contentRegex, callNumber, smsContent);
        } else {
            for (SmsMessage message : messages) {
                String callNumber = message.getOriginatingAddress();
                String smsContent = message.getMessageBody();
                if (tryFilter(callNumber, smsContent)) {
                    continue;
                }
                sb.append(String.format(Locale.CHINA, contentRegex, callNumber, smsContent));
            }
            content = sb.toString();
        }
        if (TextUtils.isEmpty(content)) {
            return;
        }
        if (Settings.is(Tags.SP_SMS_MAIL_ENABLE, false)) {
            new MailSender().send(titleRegex, content);
        }
        if (Settings.is(Tags.SP_SMS_DING_ENABLE, false)) {
            new DingDingSender().send(titleRegex, content);
        }
        if (Settings.is(Tags.SP_SMS_MAILGUN_ENABLE, false)) {
            new MailGunSender().send(titleRegex, content);
        }
        if (Settings.is(Tags.SP_SMS_TELEGRAM_ENABLE, false)) {
            new TelegramSender().send(titleRegex, content);
        }
        if (Settings.is(Tags.SP_SMS_IFTTT_WEBHOOKS_ENABLE, false)) {
            new IftttWebhooksSender().send(titleRegex, content);
        }
    }

    private boolean tryFilter(String callNumber, String content) {
        if (TextUtils.isEmpty(callNumber) || TextUtils.isEmpty(content)) {
            return true;
        }
        String senderString = App.sp.getString(Tags.SP_FILTER_VALUE_SMS_SENDER, null);
        String wordString = App.sp.getString(Tags.SP_FILTER_VALUE_SMS_KEYWORD, null);
        List<String> senderList = StringUtils.parseToList(senderString);
        List<String> wordList = StringUtils.parseToList(wordString);
        for (String number : senderList) {
            if (callNumber.contains(number)) {
                Logger.d("SMS has been filtered by sender: " + callNumber + ", " + content);
                return true;
            }
        }
        for (String word : wordList) {
            if (content.contains(word)) {
                Logger.d("SMS has been filtered by content: " + callNumber + ", " + content);
                return true;
            }
        }
        return false;
    }
}

File: app/src/main/java/com/dss886/transmis/listen/sms/SmsListener.kt
package com.dss886.transmis.listen.sms

import android.content.BroadcastReceiver
import android.content.Context
import android.content.Intent
import android.telephony.SmsMessage
import android.text.TextUtils
import com.dss886.transmis.R
import com.dss886.transmis.base.App
import com.dss886.transmis.plugin.PluginManager
import com.dss886.transmis.utils.Logger
import com.dss886.transmis.utils.Settings
import com.dss886.transmis.utils.StringUtils
import com.dss886.transmis.utils.Tags
import java.util.*

/**
 * Created by dss886 on 2017/6/29.
 */
class SmsListener : BroadcastReceiver() {

    companion object {
        private const val ACTION_SMS_RECEIVED = "android.provider.Telephony.SMS_RECEIVED"
        private const val TAG = "SmsListener"
    }

    override fun onReceive(context: Context, intent: Intent) {
        Logger.d(TAG, "SMS Received.")
        if (!Settings.`is`(Tags.SP_GLOBAL_ENABLE, false)) {
            Logger.d(TAG, "SMS Transmis has been disable!")
            return
        }
        if (ACTION_SMS_RECEIVED == intent.action) {
            val bundle = intent.extras
            if (bundle != null) {
                try {
                    val pdus = (bundle["pdus"] as Array<Any>?)!!
                    val messages = arrayOfNulls<SmsMessage>(pdus.size)
                    for (i in messages.indices) {
                        messages[i] = SmsMessage.createFromPdu(pdus[i] as ByteArray)
                    }
                    Logger.d(TAG, "Try To Notify.")
                    doNotify(messages)
                } catch (e: Exception) {
                    Logger.e(TAG, e.message ?: "")
                }
            }
        }
    }

    private fun doNotify(messages: Array<SmsMessage?>?) {
        if (messages == null || messages.isEmpty()) {
            return
        }
        val titleRegex = App.sp.getString(Tags.SP_SMS_TITLE_REGEX, null) ?: App.me().getString(R.string.sms_title_default)
        val contentRegex = App.sp.getString(Tags.SP_SMS_CONTENT_REGEX, null) ?: App.me().getString(R.string.sms_content_default)
        val content: String
        val sb = StringBuilder()
        content = if (App.sp.getBoolean(Tags.SP_SMS_MERGE_LONG_TEXT, true)) {
            for (message in messages) {
                sb.append(message?.messageBody)
            }
            val callNumber = messages[0]?.originatingAddress
            val smsContent = sb.toString()
            if (tryFilter(callNumber, smsContent)) {
                return
            }
            String.format(Locale.CHINA, contentRegex, callNumber, smsContent)
        } else {
            for (message in messages) {
                val callNumber = message!!.originatingAddress
                val smsContent = message.messageBody
                if (tryFilter(callNumber, smsContent)) {
                    continue
                }
                sb.append(String.format(Locale.CHINA, contentRegex, callNumber, smsContent))
            }
            sb.toString()
        }
        if (TextUtils.isEmpty(content)) {
            return
        }
        PluginManager.plugins
                .filter { it.isEnable() }
                .apply {
                    Logger.d(TAG, "Plugin enable count = ${this.size}: ${this.joinToString { it.getName() }}")
                }
                .forEach { plugin ->
                    plugin.doNotify(titleRegex, content)
                }
    }

    private fun tryFilter(callNumber: String?, content: String): Boolean {
        if (TextUtils.isEmpty(callNumber) || TextUtils.isEmpty(content)) {
            return true
        }
        val senderString = App.sp.getString(Tags.SP_FILTER_VALUE_SMS_SENDER, null)
        val wordString = App.sp.getString(Tags.SP_FILTER_VALUE_SMS_KEYWORD, null)
        val senderList = StringUtils.parseToList(senderString)
        val wordList = StringUtils.parseToList(wordString)
        for (number in senderList) {
            if (callNumber?.contains(number) == true) {
                Logger.d(TAG, "SMS has been filtered by sender: $callNumber, $content")
                return true
            }
        }
        for (word in wordList) {
            if (content.contains(word)) {
                Logger.d(TAG, "SMS has been filtered by content: $callNumber, $content")
                return true
            }
        }
        return false
    }

}